examples/platforms/simulation/radio.c

```diff
@@ -843,13 +850,47 @@ void radioSendAck(void)
     // Use enh-ack for 802.15.4-2015 frames
     if (otMacFrameIsVersion2015(&sReceiveFrame))
     {
+        uint8_t  linkMetricsDataLen = 0;
+        uint8_t *dataPtr            = NULL;
+
+#if OPENTHREAD_CONFIG_MLE_LINK_METRICS_ENABLE
+        uint8_t      linkMetricsData[2];
+        otMacAddress macAddress;
+
+        otEXPECT(otMacFrameGetSrcAddr(&sReceiveFrame, &macAddress) == OT_ERROR_NONE);
+
+        if (macAddress.mType == OT_MAC_ADDRESS_TYPE_SHORT)
+        {
+            linkMetricsDataLen =
+                otLinkMetricsEnhAckGetDataByShortAddress(macAddress.mAddress.mShortAddress, linkMetricsData);
+        }
+        else if (macAddress.mType == OT_MAC_ADDRESS_TYPE_EXTENDED)
+        {
+            linkMetricsDataLen =
+                otLinkMetricsEnhAckGetDataByExtAddress(&macAddress.mAddress.mExtAddress, linkMetricsData);
+        }
+
+        if (linkMetricsDataLen > 0)
+        {
+            dataPtr = linkMetricsData;
+        }
+#endif
+
+        sAckIeDataLength = generateAckIeData(dataPtr, linkMetricsDataLen);
+
         otEXPECT(otMacFrameGenerateEnhAck(&sReceiveFrame, sReceiveFrame.mInfo.mRxInfo.mAckedWithFramePending,
                                           sAckIeData, sAckIeDataLength, &sAckFrame) == OT_ERROR_NONE);
 #if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
         if (sCslPeriod > 0)
         {
             otMacFrameSetCslIe(&sAckFrame, (uint16_t)sCslPeriod, getCslPhase());
         }
+#endif
+#if OPENTHREAD_CONFIG_MLE_LINK_METRICS_ENABLE
```

Comment:
Why not moving all the previous link metrics code here? Less variables would be needed.

Reply:
I move this part into the previous code block instead. Because we need to determine the length of all IE data before generating ACK data with the IE data.

Reply:
When generating the ACK:
1. first generate header IE patterns, determine what IEs are to be included and their lengths
2. generate the ACK with the IE just generated
3. update the IE's value

Currently we use previous link metrics code to determine the header IE pattern. So we have to do it before ACK is generated. And we have to set the value of the IE after the ACK is generated.